File: internal/scan/git.go
/*
 * Copyright (C) 2026 codesafe contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 * See COPYING in the project root for the full license.
 */
// scan 包负责从 git 取 staged/working diff，并定义 commit 分类问题。
package scan

import (
	"fmt"
	"os/exec"
	"path/filepath"
	"strings"
)

// RepoRoot 返回 dir 所属仓库的顶层绝对路径；不是仓库时返回 ("", nil)。
func RepoRoot(dir string) (string, error) {
	cmd := exec.Command("git", "-C", dir, "rev-parse", "--show-toplevel")
	out, err := cmd.Output()
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			msg := strings.TrimSpace(string(ee.Stderr))
			if strings.Contains(msg, "not a git repository") {
				return "", nil
			}
			return "", fmt.Errorf("git rev-parse 失败: %s", msg)
		}
		return "", fmt.Errorf("未找到 git: %w", err)
	}
	abs, err := filepath.Abs(strings.TrimSpace(string(out)))
	if err != nil {
		return "", err
	}
	return abs, nil
}

// StagedDiff 返回 staged 区的 diff（git diff --cached）。
func StagedDiff(dir string) (string, error) {
	return gitDiff(dir, "--cached")
}

// WorktreeDiff 返回工作区相对 HEAD 的全部 diff（git diff HEAD，含暂存+未暂存）。
func WorktreeDiff(dir string) (string, error) {
	return gitDiff(dir, "HEAD")
}

// UnstagedDiff 返回未暂存部分的 diff（git diff，不含已 add 的）。
func UnstagedDiff(dir string) (string, error) {
	return gitDiff(dir)
}

// CommitDiff 返回指定 commit 的 diff（git show <rev>，含 subject 行）。
func CommitDiff(dir, rev string) (string, error) {
	cmd := exec.Command("git", "-C", dir, "show", rev, "--format=%s")
	out, err := cmd.Output()
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			return "", fmt.Errorf("git show 失败: %s", strings.TrimSpace(string(ee.Stderr)))
		}
		return "", err
	}
	return string(out), nil
}

// gitDiff 执行 git diff 并返回输出。
func gitDiff(dir string, args ...string) (string, error) {
	argv := append([]string{"-C", dir, "diff"}, args...)
	cmd := exec.Command("git", argv...)
	out, err := cmd.Output()
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			return "", fmt.Errorf("git diff 失败: %s", strings.TrimSpace(string(ee.Stderr)))
		}
		return "", err
	}
	return string(out), nil
}

File: extensions/pi/core/git.ts
/*
 * Copyright (C) 2026 codesafe contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 * See COPYING in the project root for the full license.
 */
// git.ts — minimal git access for the extension: repo root, staged/worktree
// diff, and the final `git commit`. Uses the host's shell via child_process.

import { execFile } from "node:child_process";

/** Run git and return trimmed stdout; throws on non-zero exit. */
function git(dir: string, args: string[]): Promise<string> {
	const { promise, resolve, reject } = Promise.withResolvers<string>();
	execFile("git", ["-C", dir, ...args], { maxBuffer: 64 * 1024 * 1024 }, (err, stdout, stderr) => {
		if (err) reject(new Error(stderr?.toString().trim() || String(err)));
		else resolve(stdout.toString());
	});
	return promise;
}

/** Top-level repo path for dir, or "" when not inside a git repository. */
export async function repoRoot(dir: string): Promise<string> {
	try {
		const out = await git(dir, ["rev-parse", "--show-toplevel"]);
		return out.trim();
	} catch (err) {
		if (/not a git repository/i.test(String(err))) return "";
		throw err;
	}
}

/** Staged diff (git diff --cached). */
export function stagedDiff(dir: string): Promise<string> {
	return git(dir, ["diff", "--cached"]);
}

/** Whole-worktree diff vs HEAD (staged + unstaged). */
export function worktreeDiff(dir: string): Promise<string> {
	return git(dir, ["diff", "HEAD"]);
}

/** Unstaged-only diff. */
export function unstagedDiff(dir: string): Promise<string> {
	return git(dir, ["diff"]);
}

/** Commit with the given -m arguments; returns git's combined output. */
export function commit(dir: string, messages: string[]): Promise<string> {
	const args = ["commit"];
	for (const m of messages) args.push("-m", m);
	return git(dir, args);
}
